### DISCUSSION TITLE
Setting Origin__c on LogEntry__c

### DISCUSSION THREAD

**Discussion by TKGC** [jongpie/NebulaLogger#330] [Q&A] [answered]
Is there a simple way in APEX to set the Origin__c value of a LogEntry__c record? 
I am considering the possibility of creating an abstract wrapper class (e.g. MyOrgLogHandler) to be utilized to handle logging operations within my org. In addition to decoupling, it could also abstract Logger.saveLog() from users. 

For example: MyOrgLogHandler.logInfoMessage('message');

However, wrapping the Logger calls within a custom MyOrgLogHandler class will result in the LogEntry__c.Origin__c value being set to MyOrgLogHandler - and therefore not provide that added useful peice of information in the LogEntry details. 
Is there a way to override the Origin of the LogEntry record? (e.g. setting it to the last value parsed from the stack trace) 

This may also open the possibility to abstract managing the Log__c.ParentLog__c to relate logs across transactions for a user session when applicable - I am still trying to figure out how to go about this for user session scenarios beyond Batchable and Queueable Jobs (which already have clear examples provided in the [documentation](https://github.com/jongpie/NebulaLogger#track-related-logs-in-batchable-and-queuable-jobs)). I would also be appreciative of any suggestions to this end.

Any thoughts or input is much appreciated. Thank you!

**Comment by jongpie** [comment, marked as answer]
Hi @TKGC - right now, you could manually set the origin of the entry using something like this:

```java
Logger.info('hello, world').getLogEntryEvent().OriginLocation__c = 'some value';
```

Essentially, whatever value you set on the platform event's field `OriginLocation__c` is ultimately what's stored in `LogEntry__c.OriginLocation__c` (which is used in the formula field `LogEntry__c.Origin__c`). I don't think this approach is ideal, as you would need to add additional logic to fully handle parsing a stack trace (which is something Nebula Logger already does internally), but it's something you could implement today. However, I already have [some internal logic built to skip some classes](https://github.com/jongpie/NebulaLogger/blob/31e510a602c06b8cb1fb13237a92753fab3b3826/nebula-logger/core/main/logger-engine/classes/LogEntryEventBuilder.cls#L780) when setting the field `OriginLocation__c` - I'll look into the possibility of making this list of classes configurable using the custom metadata type `LoggerParameter__mdt`, I think that would be the better long-term solution for this kind of situation.

A couple of follow up questions for you though, as I'm curious about your use case:

- Other than decoupling the rest of your code from Nebula Logger, what's the advantage of introducing your `MyOrgLogHandler` class? Offhand, it sounds like potentially an unnecessary extra layer, but I'd love to hear more about why you're considering using this approach (instead of just calling `Logger` directly throughout your codebase)
- Can you clarify what you mean by "abstract `Logger.saveLog()` from users?"
- Can you clarify what you mean by "abstract managing the `Log__c.ParentLog__c` to relate logs across transactions for a user session when applicable?"

**Reply by TKGC** [reply]
Thank you @jongpie for the very quick response!! 

I appreciate the manual work around for setting the `OriginLocation__c`, and will look into your logic for parsing the stack trace. 


For some context - I'm exploring potential options or best practices for our devs to adopt/implement Nebula Logger across our Org as we start to become more familiar with it and its capabilities. In the end, we certainly may implement calling Logger directly throughout the code base as you mentioned. But since we're just getting started, figured now would be the time to explore abstraction options to determine if there are viable benefits. 
Hopefully I can gather some of my thoughts so far and make enough sense to add some clarification/context to my original questions.

In response to your questions:  
1. The main reason would be decoupling. Beyond that, the wrapper class could automatically call saveLog(), so developers won't need to remember to add that line of code or [think about where to best add it](https://github.com/jongpie/NebulaLogger/discussions/244). Another option could be to abstract any special formatting (json, xml) or addition of custom attributes to logged messages to better accommodate any log analysis platforms we decide to utilize downstream in exporting the logs (Splunk, NewRelic, etc.)  

2. As far as "abstracting managing the Log__c.ParentLog__c to relate logs across transactions for a user session": Since this is such a powerful concept that you've successfully implemented functionality for, the train of thought might be to implement something along the lines of what's described in [this video by Matt Gerry](https://www.youtube.com/watch?v=AB4RhfRVvyk&t=1401s). Not sure if this session cache example is viable with code run in the context of an integration user - will have to test that out. Is this concept something you've thought about or can provide any insight on other potential implementations? 




BTW, thank you for taking the time and sharing your expertise in creating such a fantastic tool to share with the community.

**Reply by jongpie** [reply]
@TKGC that all makes sense, thanks for the explanation! Yeah, I think especially since you're just now starting to integrate Nebula Logger, decoupling to start isn't a bad idea if that's what you prefer. As you continue testing out Nebula Logger, just let me know if you have any questions along the way. 

And just a few follow up thoughts to your last comment:

- Decoupling makes sense to start, but if/when you decide that you definitely are going to move forward with using Nebula Logger, then I think you'll want to eventually started calling the `Logger` class directly. If you continue to use your own wrapper class, then you'll end up in 1 of 2 situations:
  - Option 1: your wrapper class only has a few public methods available for some of Nebula Logger's features, so the rest of your codebase can't leverage all of Nebula Logger's features
  - Option 2: your wrapper class has equivalent public methods for most/all of Nebula Logger's methods, which will become a headache for you to maintain
- I would be cautious about automatically calling `Logger.saveLog()` in your wrapper class - each call to save counts towards transaction limits for publishing platform events, so you want to be careful about how frequently you call the save method. There are definitely times where it's ok to auto-call it - auto-saving within a trigger handler framework is one of my favorite places to do so, but you just want to be strategic about when/where you auto-save. In general though, developers _need_ to think about where to put it and remember to call it.
- For "abstracting managing the Log__c.ParentLog__c to relate logs across transactions for a user session," I'm not sure that I fully understand what you're suggesting to implement, but I'd love to know more about what you have in mind. Let me know if you can share some more details on this part.

**Reply by TKGC** [reply]
Thanks again @jongpie for taking the time to reply so quickly and thoroughly to community questions. We actually have recently decided to expand the implementation of Nebula Logger in our org, and your responsiveness and input has solidified my confidence in that decision.

The wrapper class situations you outlined make sense and are good to think about. 
And your points about making it a practice for developers to consciously think about when and where to call `saveLog()` are well taken.  

An interesting thought just came to mind as I read your first point again, I wonder how the dynamic would change if the Logger class was defined as `virtual` and a wrapper class extended it? 
The wrapper class could add custom method functionality while still exposing all other Logger methods directly. However, I think it still wouldn't achieve the decoupling benefit without the maintenance headache you mentioned and would still require any custom methods in the wrapper to correctly set the `LogEntry__c.OriginLocation__c` value...

To your question about more details on "abstracting managing the `Log__c.ParentLog__c`": Were you able to view [this youtube video I linked](https://youtu.be/AB4RhfRVvyk?t=1386) to? The thought would be to implement something along those lines.

**Reply by jongpie** [reply]
Hi @TKGC - unfortunately marking the `Logger` class as virtual would not have any impact, due to all of its methods being `static`, so it's not feasible to extend the class. You could still have your own wrapper class with additional methods, but it just wouldn't be done via extending the `Logger` class.

Regarding your original question about setting the `Origin__c` field on `LogEntry__c`, I've created #335 to track that enhancement - I'm going to experiment with making it possible with either configuration (stored in `LoggerParameter__mdt`) as well as via a new method on either `Logger` or `LogEntryBuilder` that can be called via Apex. I'm hoping I can get it implemented within the next 1-2 weeks, I'll keep you updated on the status.

And lastly, regarding Matt Gerry's youtube video - the main concern I have is that it relies on Platform Cache, which is not available in all Salesforce orgs, so I've been hesitant to use it within Nebula Logger. But I'll do some prototyping (probably in a few months) to see if there is another way to accomplish the same functionality.

**Reply by TKGC** [reply]
Thanks @jongpie! I appreciate the enhancement submission as well as your willingness to investigate and prototype options to allow for mantaining `Log__c.ParentLog__c` in the use case described. 
Please let me know if I can provide any additional information that might assist. 


On the other topic of marking Logger as virtual: I think I see your point about static methods not being inheritable. I was playing around with some examples (see below), and I guess the option a child class extending Logger could yield, would be 'hiding' Logger static methods of the same signature if the static method is called from the child class, while continuing to expose parent static methods not redefined in the child class. Which could yield some interesting options. 
_Full disclosure: I am a relatively new to apex, so please excuse any ignorance in this space._ 

Lastly, I know we have three topics going in this discussion thread, and I very much appreciate you entertaining them, as I am learning from it.


```
public virtual with sharing class VirtualClass {
    
    public VirtualClass(){
        
    }
    
    public static void method1(){
        system.debug('>>> VirtualClass.method1'); 
    }
    
    public static void method2(){
        system.debug('>>> VirtualClass.method2');
    }
 
}

public with sharing class WrapperClass extends VirtualClass {
    
    public static void method1(){
        system.debug('>>> WrapperClass.method1');
    }
    
    public static void wrapperMethod4(){
        system.debug('>>> WrapperClass.wrapperMethod4');
    }
    
}

//Results:
VirtualClass.method1(); 		// >>> VirtualClass.method1 
WrapperClass.method1(); 		// >>> WrapperClass.method1 
WrapperClass.method2(); 		// >>> VirtualClass.method2 
WrapperClass.wrapperMethod4(); 	        // >>> WrapperClass.wrapperMethod4 

```

**Reply by jongpie** [reply]
@TKGC just another update on this discussion - [my PR for v4.8.0](https://github.com/jongpie/NebulaLogger/pull/343) will add a new method to allow you to specify your own classes to ignore when logger is populating the log entry's origin. Once `v4.8.0` is released, you'll be able to have your class excluded by calling the new method `Logger.ignoreOrigin()`. For example:

```java
Logger.ignoreOrigin(SomeFakeApex.class);
Logger.info('hello, world').getLogEntryEvent();
```

Depending on if your class is using static methods or instance methods (or both), you can call the method in 1 of 2 ways to have your class ignored:

```java
public class SomeFakeApex {
    static {
        // Call Logger.ignoreOrigin() in your class's static initializer to automatically have this apply to all of your static methods
        Logger.ignoreOrigin(SomeFakeApex.class);
    }

    @AuraEnabled
    public static void doSomething() {
        // Since this is a static method, the static initialization above handles calling Logger.ignoreOrigin() - no need to call it again within this method
        Logger.info('logging something from a static method!');
        Logger.saveLog();
    }

    public SomeFakeApex() {
        // Call Logger.ignoreOrigin() in your class's constructor(s) to automatically have this apply to all of your instance methods
        Logger.ignoreOrigin(SomeFakeApex.class);
    }

    public void doSomethingElse() {
        // Since this is an instance method, the class's constructor above handles calling Logger.ignoreOrigin() - no need to call it again within this method
        Logger.info('logging something from an instancemethod!');
        Logger.saveLog();
    }
}
```

With the above class deployed, running this anonymous apex script would result in 2 log entries, both with an `OriginLocation__c` of 'AnonymousBlock' (instead of using `SomeFakeApex`):

```java
SomeFakeApex.doSomething();
new SomeFakeApex().doSomethingElse();
``` 

The new `v4.8.0` release will hopefully be available later this week!

**Reply by TKGC** [reply]
Hi @jongpie, that's fantastic news!!
Truly appreciate such a quick turn around on this enhancement - can't wait to try it out when it becomes available!

**Reply by jongpie** [reply]
@TKGC thanks for all of your feedback! I'm still trying to wrap up some unrelated testing for the new release, but this feature was actually pretty easy to add, and I'm really excited about it. If you or your team have other feedback, just let me know!

**Reply by TKGC** [reply]
Will do, @jongpie! We'll utilize it in short order. 
I have no doubt others in the community will also find this feature useful!

**Reply by jongpie** [reply]
@TKGC this is now finally available in the latest release, [v4.8.0](https://github.com/jongpie/NebulaLogger/releases/tag/v4.8.0)! Once you've had a chance to use it, let me know if you have any feedback

**Reply by TKGC** [reply]
Thanks @jongpie!! Will do.

**Reply by kevlar8** [reply]
Hi @jongpie, this feature looks exactly like what I need as I have a requirement to decouple our logging. 

I tried it out but I see that my requirement is slightly different in that I have two levels of classes before I call the Nebula Logger class

i.e. MyClass -> MainLogger -> NebulaWrapper -> Logger

So I tried
        Logger.ignoreOrigin(MainLogger.class);
        Logger.ignoreOrigin(NebulaWrapper.class);

Which does remove the named classes from the stack trace which is great.

But what I am seeing in the origin field is the value **Apex.External entry point** and I am also seeing **External entry point** as the first line of the stack trace and not MyClass.

Is there any way to get around this?

**Reply by jongpie** [reply]
Hi @kevlar8, that's very strange, I've not previously seen the origin field/stack trace populate with 'External entry point'. Could you try adding this line to your method in `MyClass`? This will print out the original, unmodified stack trace - it may contain some details that you don't want to share publicly, but can you see if this also includes 'External entry point' in the output?

```java
public class MyClass {
  public static void someMethod() {
     // You can add this codeto your existing method, and then see what it outputs in developer console or the LogEntry__c object
     Logger.debug('>> The original stack trace: ' + new DmlException().getStackTraceString());
     Logger.saveLog();
  }
}
```

**Reply by kevlar8** [reply]
Hi @jongpie, thanks for investigating this. 

I added the above code to my init method and got Apex.MyClass.init set in the Origin and could not see "Apex.External entry point" in the stack trace. 

One other thing i found out is that if I only have one level of calls before the wrapper 
        i.e. MyClass -> NebulaWrapper -> Logger

I do not get Apex.External entry point in the stack trace. It seems to be that extra level of call that causes the issue for me.

**Reply by jongpie** [reply]
That's really interesting, thanks for sharing @kevlar8 - I've not seen that issue before, but I'll try to setup some similar classes to see if I can recreate it. Offhand, I don't know of a solution, but I'll do some research to see if anything can be done to better handle this scenario.
